File: src/django_grid_view/search/column_scope.py
"""Column label scoping for toolbar ``q`` (``лікарів:10..20``)."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import TypedDict

__all__ = [
    "ColumnSearchMeta",
    "cells_for_scope",
    "column_keys_for_hint",
    "normalize_column_label",
    "parse_scoped_or_group",
]


class ColumnSearchMeta(TypedDict):
    key: str
    label: str


def normalize_column_label(text: str) -> str:
    return str(text or "").casefold().replace(" ", "")


def column_keys_for_hint(hint: str, columns: Sequence[ColumnSearchMeta]) -> list[str]:
    """Match visible column labels by substring (space/case insensitive)."""
    needle = normalize_column_label(hint)
    if not needle:
        return []
    keys: list[str] = []
    for col in columns:
        label = normalize_column_label(col["label"])
        if needle in label or label.startswith(needle):
            keys.append(col["key"])
    return keys


def parse_scoped_term(
    term: str,
    columns: Sequence[ColumnSearchMeta],
) -> tuple[str | None, str]:
    """If *term* is ``label:…`` and *label* matches a column, return scope hint + inner term."""
    raw = str(term or "").strip()
    if not raw or ":" not in raw:
        return None, raw
    idx = raw.find(":")
    hint = raw[:idx].strip()
    rest = raw[idx + 1 :].strip()
    if not hint or not rest:
        return None, raw
    if column_keys_for_hint(hint, columns):
        return hint, rest
    return None, raw


def parse_scoped_or_group(
    group: str,
    columns: Sequence[ColumnSearchMeta],
) -> tuple[str | None, str]:
    """Legacy alias — scope is resolved per AND term, not per OR group."""
    return parse_scoped_term(group, columns)


def cells_for_scope(
    hint: str | None,
    cells_by_key: Mapping[str, str],
    columns: Sequence[ColumnSearchMeta],
    *,
    all_cells: Sequence[str],
) -> list[str]:
    if not hint:
        return list(all_cells)
    keys = column_keys_for_hint(hint, columns)
    if not keys:
        return list(all_cells)
    scoped: list[str] = []
    for key in keys:
        val = cells_by_key.get(key)
        if val:
            scoped.append(val)
    return scoped

File: frontend/src/grid-view/search/column-scope.ts
/** Column label scoping for toolbar ``q`` — mirror of grid_view_spec.search.column_scope */

export interface ColumnSearchMeta {
  key: string;
  label: string;
}

export function normalizeColumnLabel(text: unknown): string {
  return String(text ?? "")
    .toLowerCase()
    .replace(/\s+/g, "");
}

export function columnKeysForHint(
  hint: string,
  columns: readonly ColumnSearchMeta[]
): string[] {
  const needle = normalizeColumnLabel(hint);
  if (!needle) return [];
  const keys: string[] = [];
  for (const col of columns) {
    const label = normalizeColumnLabel(col.label);
    if (label.includes(needle) || label.startsWith(needle)) {
      keys.push(col.key);
    }
  }
  return keys;
}

export function parseScopedTerm(
  term: string,
  columns: readonly ColumnSearchMeta[]
): { scope: string | null; inner: string } {
  const raw = String(term ?? "").trim();
  if (!raw || raw.indexOf(":") < 0) return { scope: null, inner: raw };
  const idx = raw.indexOf(":");
  const hint = raw.slice(0, idx).trim();
  const inner = raw.slice(idx + 1).trim();
  if (!hint || !inner) return { scope: null, inner: raw };
  if (columnKeysForHint(hint, columns).length) return { scope: hint, inner };
  return { scope: null, inner: raw };
}

/** @deprecated Scope is resolved per AND term. */
export function parseScopedOrGroup(
  group: string,
  columns: readonly ColumnSearchMeta[]
): { scope: string | null; rest: string } {
  const { scope, inner } = parseScopedTerm(group, columns);
  return { scope, rest: inner };
}

export function cellsForScope(
  hint: string | null,
  cellsByKey: Readonly<Record<string, string>>,
  columns: readonly ColumnSearchMeta[],
  allCells: readonly string[]
): string[] {
  if (!hint) return [...allCells];
  const keys = columnKeysForHint(hint, columns);
  if (!keys.length) return [...allCells];
  const scoped: string[] = [];
  for (const key of keys) {
    const val = cellsByKey[key];
    if (val) scoped.push(val);
  }
  return scoped;
}
